kernel/sem.c

```diff
@@ -98,18 +113,165 @@ int z_vrfy_k_sem_init(struct k_sem *sem, unsigned int initial_count,
 static inline void handle_poll_events(struct k_sem *sem)
 {
 #ifdef CONFIG_POLL
+#ifdef CONFIG_SEM_IMPL_ATOMIC
+	k_spinlock_key_t key = k_spin_lock(&poll_lock);
+
+#endif
 	z_handle_obj_poll_events(&sem->poll_events, K_POLL_STATE_SEM_AVAILABLE);
+#ifdef CONFIG_SEM_IMPL_ATOMIC
+	k_spin_unlock(&poll_lock, key);
+#endif
 #else
 	ARG_UNUSED(sem);
 #endif
 }
 
-void z_impl_k_sem_give(struct k_sem *sem)
+#ifdef CONFIG_SEM_IMPL_ATOMIC
+/*
+ * Quick overview of this semaphore implementation:
+ * `count` is always eventually equal to `limit`, minus the number
+ * of outstanding `sem_take`s, minus the number of waiting threads
+ * in `wait_q`.
+ *
+ * Once a thread has incremented or decremented `count`, it is
+ * responsible for ensuring the above is promptly restored.
+ *
+ */
+
+static void sem_give(struct k_sem *sem)
+{
+	atomic_t count;
+	atomic_t limit_minus_one = sem->limit - 1;
+	unsigned int key = arch_irq_lock();
+
+	do {
+		count = atomic_get(&sem->count);
+	} while (!atomic_cas(&sem->count, count, MIN(count, limit_minus_one) + 1));
+
+	while (count < 0) {
+		struct k_thread *thread = z_unpend_first_thread(&sem->wait_q);
```

Comment:
I see (I think) why this is safe vs. SMP races with itself.  But how does this work vs. k_sem_take().  Seems like another thread might have come along and pended itself after this unpend call?  I believe this probably works, but I'm having trouble working it out from scratch.  Maybe a quick description of the players and their possible implicit states would be helpful?

Reply:
> how does this work vs. k_sem_take()

Very subtly...

>  Seems like another thread might have come along and pended itself after this unpend call?

Yep, that's fine. The only way the other thread pended itself is if it went through k_sem_take, in which case `count` was updated and will be picked up by the reload of sem->count in the loop...

...assuming I am holding this entire complex thing in my head correctly, which may be a bad assumption.

*****

I was going to model this using TLA+, but there's no good way to model a weak memory model there... I'll do up a model anyway.